Bluetooth: Controller: Updated ISO-AL time wrapping and restructured unit test

niag-Oticon opened the pull request:
-- Handle time calculations across the point at which the controller's time information from the controller's clock will wrap (in the
   microsecond range)
-- Use wrapped time in all ISO-AL time calculations
-- Included TX unit tests
-- Re-factored unit tests to use FFF

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -39,6 +42,14 @@
+ */
+#define ISOAL_TIME_WRAPPING_POINT         (HAL_TICKER_TICKS_TO_US(HAL_TICKER_CNTR_MASK))
+#define ISOAL_TIME_MID_POINT              (ISOAL_TIME_WRAPPING_POINT / 2)
+#define ISOAL_TIME_SPAN_FULL              (ISOAL_TIME_WRAPPING_POINT + 1)
+#define ISOAL_TIME_SPAN_HALF              (ISOAL_TIME_SPAN_FULL / 2)
```
Why is `SPAN_HALF` not `ISOAL_TIME_SPAN_FULL / 2`?

niag-Oticon replied:
Sorry but I don't think that I understood the question. It is ISOAL_TIME_SPAN_FULL / 2.

Thalley replied:
Yes, sorry, I misread :|

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -89,6 +100,22 @@ isoal_status_t isoal_reset(void)
 }
 
+/**
+ * @breif Wraps given time within the range of 0 to ISOAL_TIME_WRAPPING_POINT
```
`@brief`

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -39,6 +42,14 @@
+/* Defined the wrapping point and mid point in the range of time input values,
+ * which depend on range of the controller's clock in microseconds.
+ */
+#define ISOAL_TIME_WRAPPING_POINT         (HAL_TICKER_TICKS_TO_US(HAL_TICKER_CNTR_MASK))
+#define ISOAL_TIME_MID_POINT              (ISOAL_TIME_WRAPPING_POINT / 2)
+#define ISOAL_TIME_SPAN_FULL              (ISOAL_TIME_WRAPPING_POINT + 1)
+#define ISOAL_TIME_SPAN_HALF              (ISOAL_TIME_SPAN_FULL / 2)
```
IMO it's typically useful to denote the time unit in macros like these (s, ms, us)

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -89,6 +100,22 @@ isoal_status_t isoal_reset(void)
+ * @param  time_diff Time difference (signed)
+ * @return           Wrapped time after difference
+ */
+static uint32_t isoal_get_wrapped_time(uint32_t time_now, int32_t time_diff)
+{
+	LL_ASSERT(time_now <= ISOAL_TIME_WRAPPING_POINT);
+
+	uint32_t result = ((uint64_t)time_now + ISOAL_TIME_SPAN_FULL + time_diff) %
+				((uint64_t)ISOAL_TIME_SPAN_FULL);
+
+	return result;
+}
```
Ditto, it's IMO useful to denote the time units

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -1088,6 +1114,50 @@ isoal_status_t isoal_rx_pdu_recombine(isoal_sink_handle_t sink_hdl,
 
 #if defined(CONFIG_BT_CTLR_ADV_ISO) || defined(CONFIG_BT_CTLR_CONN_ISO)
+/**
+ * @breif Check if a time difference calculation is valid and return the difference.
```
`@brief`

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -1088,6 +1114,50 @@ isoal_status_t isoal_rx_pdu_recombine(isoal_sink_handle_t sink_hdl,
+ * @breif Check if a time difference calculation is valid and return the difference.
+ * @param  time_before Subtrahend
+ * @param  time_after  Minuend
+ * @param  result      Difference if valid
+ * @return             Validity
```
Please provide information on when a difference is valid (as far as I can tell, if `time_before` is after `time_after`, then it is invalid)

Thalley commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -1814,18 +1885,27 @@ static isoal_status_t isoal_tx_framed_produce(struct isoal_source *source,
+					actual_grp_ref_point, &time_diff);
+		LL_ASSERT(time_diff_valid);
+		LL_ASSERT(time_diff > 0);
+		/* Time difference should be less than the maximum possible
+		 * time-offset of 24-bits.
+		 */
+		LL_ASSERT(time_diff <= 0x00FFFFFF);
```
`should` and `assert` doesn't seem to go hand in hand. Should this have been a warning? Or should it have been a `shall`?

niag-Oticon replied:
If the time difference exceeds 24-bits there is something very wrong in either the SDU's timestamp or its target event's reference point. So should / shall ---> must and I would say that the assert is valid in this case.

Thalley reviewed (dismissed):
A few comments for the ISO time wrapping. Won't review the unit tests

mtpr-ot commented on tests/bluetooth/ctrl_isoal/testcase.yaml:
```diff
@@ -1,44 +1,82 @@
+      - tx_framed_1_sdu_1_frag_pdu_alloc_err
+      - tx_framed_1_sdu_1_frag_pdu_emit_err
+      - tx_framed_2_sdu_1_frag_pdu_timeout
+      - tx_framed_cis_fra_per_bv07c-
```
Why is there a "-" at the end? Doesn't this exclude the test case?

niag-Oticon replied:
Good question! I didn't really consider it as I followed the existing format. 
But Google tells me that it's part of the YAML markup language denoting members of a list. In this case `testcases`.

mtpr-ot replied:
The trailing "-"? I don't think so

niag-Oticon replied:
Nice catch. Didn't notice.

niag-Oticon replied:
Removed.

mtpr-ot reviewed (dismissed):
I have not reviewed the unit tests in detail, but overall structure, coverage and quality seems very good. I approve on that basis.

niag-Oticon committed:
tests: bluetooth: controller: restructured and update ISO-AL unit tests

-- Included TX tests
-- Re-factored to use FFF

Signed-off-by: Nirosharn Amarasinghe <niag@demant.com>

mtpr-ot reviewed (dismissed):
LGTM

cvinayak commented on subsys/bluetooth/controller/ll_sw/isoal.c:
```diff
@@ -27,6 +29,7 @@
 #define LOG_MODULE_NAME bt_ctlr_isoal
 #include "common/log.h"
 #include "hal/debug.h"
+#include "hal/ticker.h"
```
Move this between like this:
```c
#include "util/memq.h"

#include "hal/ticker.h"

#include "pdu.h"
```

cvinayak reviewed (changes requested):
Only reviewed isoal.c. Minor nitpick otherwise looks good.

niag-Oticon committed:
Bluetooth: controller: Time wrapping point used in ISO-AL calculations

-- Handle time calculations across the point at which the controller's
   time information from the controller's clock will wrap (in the
   microsecond range)
-- Use wrapped time in all ISO-AL time calculations

Signed-off-by: Nirosharn Amarasinghe <niag@demant.com>

mtpr-ot reviewed (approved):
LGTM
